File: imagegen/imagegen-cli/src/commands/generate.ts
import { mkdir, writeFile } from "node:fs/promises";
import { dirname, resolve } from "node:path";
import { GoogleGenAI } from "@google/genai";
import { defineCommand } from "citty";
import { InputError, type InputImage, readInputImage } from "../lib/inputs.ts";
import { getCapabilities } from "../lib/models.ts";
import { jsonError, jsonOk } from "../lib/output.ts";

const DEFAULT_MODEL = "gemini-3.1-flash-image-preview";

export const generateCommand = defineCommand({
	meta: {
		name: "generate",
		description:
			"Generate or edit an image with Gemini. Pass --image one or more times to edit, restyle, or compose existing images.",
	},
	args: {
		prompt: {
			type: "positional",
			description:
				"Describe the desired image. Provide intent and context; let the model handle creative details.",
			required: true,
		},
		output: {
			type: "string",
			alias: "o",
			description: "Output file path (default: generated_{timestamp}.png)",
		},
		image: {
			type: "string",
			alias: "i",
			description:
				"Input image path (PNG/JPEG/WEBP/GIF). Repeat for multiple references.",
		},
		"aspect-ratio": {
			type: "string",
			alias: "a",
			description: "Aspect ratio (model-validated, e.g. 1:1, 16:9, 21:9)",
		},
		size: {
			type: "string",
			alias: "s",
			description: "Image size: 512, 1K, 2K, 4K (model-dependent)",
		},
		thinking: {
			type: "string",
			alias: "t",
			description:
				"Thinking level: minimal | high. Only honored by gemini-3.1-flash-image-preview.",
		},
		model: {
			type: "string",
			alias: "m",
			description: "Model id (default: gemini-3.1-flash-image-preview)",
			default: DEFAULT_MODEL,
		},
		"negative-prompt": {
			type: "string",
			alias: "n",
			description:
				"Things to exclude. Prefer rewriting the prompt positively when possible.",
		},
	},
	async run({ args }) {
		const apiKey = process.env.GEMINI_API_KEY;
		if (!apiKey) {
			jsonError(
				"GEMINI_API_KEY environment variable is not set",
				"API_KEY_MISSING",
				"Get your API key at https://aistudio.google.com/apikey and set it: export GEMINI_API_KEY=your_key",
			);
		}

		const prompt = args.prompt;
		const aspectRatio = args["aspect-ratio"] || "";
		const size = args.size || "";
		const thinking = (args.thinking || "").toLowerCase();
		const model = args.model || DEFAULT_MODEL;
		const negativePrompt = args["negative-prompt"] || "";

		// citty yields an array when a flag repeats; coerce to string[].
		const rawImage = args.image as string | string[] | undefined;
		const imagePaths: string[] = rawImage
			? Array.isArray(rawImage)
				? rawImage
				: [rawImage]
			: [];

		// --- Capability-driven validation -----------------------------------
		const caps = getCapabilities(model);
		if (!caps) {
			console.error(
				`[imagegen] Warning: unknown model '${model}'. Skipping capability validation; the API may reject the call.`,
			);
		}

		if (caps && aspectRatio && !caps.aspectRatios.includes(aspectRatio)) {
			jsonError(
				`Invalid aspect ratio '${aspectRatio}' for model '${model}'. Valid: ${caps.aspectRatios.join(", ")}`,
				"INVALID_PARAMS",
			);
		}

		if (caps && size) {
			if (caps.imageSizes === null) {
				jsonError(
					`Model '${model}' does not accept --size. Omit the flag or switch to a model that supports it (e.g. gemini-3.1-flash-image-preview).`,
					"INVALID_PARAMS",
				);
			} else if (!caps.imageSizes.includes(size)) {
				jsonError(
					`Invalid size '${size}' for model '${model}'. Valid: ${caps.imageSizes.join(", ")}`,
					"INVALID_PARAMS",
				);
			}
		}

		if (caps && thinking) {
			if (caps.thinkingLevels === null) {
				jsonError(
					`Model '${model}' does not accept --thinking. Omit the flag or use gemini-3.1-flash-image-preview.`,
					"INVALID_PARAMS",
				);
			} else if (!caps.thinkingLevels.includes(thinking)) {
				jsonError(
					`Invalid thinking level '${thinking}' for model '${model}'. Valid: ${caps.thinkingLevels.join(", ")}`,
					"INVALID_PARAMS",
				);
			}
		}

		if (caps && imagePaths.length > caps.maxInputImages) {
			jsonError(
				`Too many input images (${imagePaths.length}) for model '${model}'. Max: ${caps.maxInputImages}.`,
				"INVALID_PARAMS",
			);
		}

		// --- Load input images ----------------------------------------------
		let inputImages: InputImage[] = [];
		try {
			inputImages = await Promise.all(imagePaths.map(readInputImage));
		} catch (err) {
			if (err instanceof InputError) {
				jsonError(`${err.message} (${err.path})`, "INPUT_ERROR");
			}
			throw err;
		}

		// --- Build request --------------------------------------------------
		const fullPrompt = negativePrompt
			? `${prompt}. Do not include: ${negativePrompt}.`
			: prompt;

		const parts: Array<
			{ text: string } | { inlineData: { mimeType: string; data: string } }
		> = [{ text: fullPrompt }];
		for (const img of inputImages) {
			parts.push({ inlineData: { mimeType: img.mimeType, data: img.data } });
		}

		const imageConfig: Record<string, string> = {};
		if (aspectRatio) imageConfig.aspectRatio = aspectRatio;
		if (size) imageConfig.imageSize = size;

		const config: Record<string, unknown> = {
			responseModalities: ["IMAGE"],
		};
		if (Object.keys(imageConfig).length > 0) {
			config.imageConfig = imageConfig;
		}
		if (thinking) {
			// SDK enum is upper-case (MINIMAL | HIGH). Normalize.
			config.thinkingConfig = { thinkingLevel: thinking.toUpperCase() };
		}

		const outputPath = resolve(args.output || `generated_${Date.now()}.png`);

		// --- Call API -------------------------------------------------------
		try {
			const ai = new GoogleGenAI({ apiKey });
			const response = await ai.models.generateContent({
				model,
				contents: [{ role: "user", parts }],
				config,
			});

			const responseParts = response.candidates?.[0]?.content?.parts;
			if (!responseParts || responseParts.length === 0) {
				jsonError("No content in API response", "API_ERROR");
			}

			// Pick the final image: prefer the last non-thought image part;
			// fall back to the last image part if all are thoughts.
			const imageParts = responseParts.filter((p) =>
				p.inlineData?.mimeType?.startsWith("image/"),
			);
			if (imageParts.length === 0) {
				jsonError("API response did not contain an image", "API_ERROR");
			}
			const nonThought = imageParts.filter((p) => !p.thought);
			const finalPart = (nonThought.length > 0 ? nonThought : imageParts).at(
				-1,
			);
			const inline = finalPart?.inlineData;
			if (!inline?.data || !inline.mimeType) {
				jsonError("Image data missing from API response", "API_ERROR");
			}

			const buffer = Buffer.from(inline.data, "base64");
			try {
				await mkdir(dirname(outputPath), { recursive: true });
				await writeFile(outputPath, buffer);
			} catch (err) {
				const reason = err instanceof Error ? err.message : "unknown error";
				jsonError(`Could not write output file: ${reason}`, "OUTPUT_ERROR");
			}

			const result: Record<string, unknown> = {
				file: outputPath,
				mimeType: inline.mimeType,
				size: buffer.length,
				prompt,
				model,
			};
			if (aspectRatio) result.aspectRatio = aspectRatio;
			if (size) result.imageSize = size;
			if (thinking) result.thinkingLevel = thinking;
			if (inputImages.length > 0) {
				result.inputImages = inputImages.map((i) => i.absolutePath);
			}
			jsonOk(result);
		} catch (err) {
			const message = err instanceof Error ? err.message : "Unknown API error";
			jsonError(message, "API_ERROR");
		}
	},
});

File: imagegen/imagegen-cli/src/imagegen_cli/commands/generate.py
"""``imagegen generate`` — text-to-image and image-editing via Google Gemini."""

from __future__ import annotations

import sys
import time
from pathlib import Path
from typing import Any

import click

from ..api_key import resolve_api_key
from ..inputs import InputError, InputImage, read_input_image
from ..models import DEFAULT_MODEL, get_capabilities
from ..output import json_error, json_ok

_API_KEY_MISSING_HINT = (
    "Set GEMINI_API_KEY env var, or write the key to '.gemini-key' in the "
    "project root (gitignore it) or '~/.gemini-key' for machine-wide use. "
    "Get a key at https://aistudio.google.com/apikey."
)


@click.command(
    "generate",
    short_help="Generate or edit an image with Gemini.",
    help=(
        "Generate or edit an image with Gemini. Pass --image one or more times "
        "to edit, restyle, or compose existing images."
    ),
)
@click.argument("prompt")
@click.option(
    "-o",
    "--output",
    "output",
    default=None,
    help="Output file path (default: generated_{timestamp}.png)",
)
@click.option(
    "-i",
    "--image",
    "image_paths",
    multiple=True,
    help="Input image path (PNG/JPEG/WEBP/GIF). Repeat for multiple references.",
)
@click.option(
    "-a",
    "--aspect-ratio",
    "aspect_ratio",
    default="",
    help="Aspect ratio (model-validated, e.g. 1:1, 16:9, 21:9).",
)
@click.option(
    "-s",
    "--size",
    "size",
    default="",
    help="Image size: 512, 1K, 2K, 4K (model-dependent).",
)
@click.option(
    "-t",
    "--thinking",
    "thinking",
    default="",
    help="Thinking level: minimal | high. Only honored by gemini-3.1-flash-image-preview.",
)
@click.option(
    "-m",
    "--model",
    "model",
    default=DEFAULT_MODEL,
    show_default=True,
    help="Model id.",
)
@click.option(
    "-n",
    "--negative-prompt",
    "negative_prompt",
    default="",
    help="Things to exclude. Prefer rewriting the prompt positively when possible.",
)
def cmd(
    prompt: str,
    output: str | None,
    image_paths: tuple[str, ...],
    aspect_ratio: str,
    size: str,
    thinking: str,
    model: str,
    negative_prompt: str,
) -> None:
    # --- API key --------------------------------------------------------
    resolved = resolve_api_key()
    if resolved is None:
        json_error(
            "Gemini API key not found",
            code="API_KEY_MISSING",
            hint=_API_KEY_MISSING_HINT,
        )

    # Normalize.
    thinking = thinking.lower()

    # --- Capability-driven validation -----------------------------------
    caps = get_capabilities(model)
    if caps is None:
        # Match the TS warning verbatim; goes to stderr so stdout stays clean.
        print(
            f"[imagegen] Warning: unknown model '{model}'. "
            "Skipping capability validation; the API may reject the call.",
            file=sys.stderr,
        )

    if caps is not None and aspect_ratio and aspect_ratio not in caps.aspect_ratios:
        json_error(
            f"Invalid aspect ratio '{aspect_ratio}' for model '{model}'. "
            f"Valid: {', '.join(caps.aspect_ratios)}",
            code="INVALID_PARAMS",
        )

    if caps is not None and size:
        if caps.image_sizes is None:
            json_error(
                f"Model '{model}' does not accept --size. Omit the flag or "
                "switch to a model that supports it "
                "(e.g. gemini-3.1-flash-image-preview).",
                code="INVALID_PARAMS",
            )
        elif size not in caps.image_sizes:
            json_error(
                f"Invalid size '{size}' for model '{model}'. "
                f"Valid: {', '.join(caps.image_sizes)}",
                code="INVALID_PARAMS",
            )

    if caps is not None and thinking:
        if caps.thinking_levels is None:
            json_error(
                f"Model '{model}' does not accept --thinking. Omit the flag "
                "or use gemini-3.1-flash-image-preview.",
                code="INVALID_PARAMS",
            )
        elif thinking not in caps.thinking_levels:
            json_error(
                f"Invalid thinking level '{thinking}' for model '{model}'. "
                f"Valid: {', '.join(caps.thinking_levels)}",
                code="INVALID_PARAMS",
            )

    if caps is not None and len(image_paths) > caps.max_input_images:
        json_error(
            f"Too many input images ({len(image_paths)}) for model "
            f"'{model}'. Max: {caps.max_input_images}.",
            code="INVALID_PARAMS",
        )

    # --- Load input images ----------------------------------------------
    input_images: list[InputImage] = []
    try:
        for raw_path in image_paths:
            input_images.append(read_input_image(raw_path))
    except InputError as err:
        json_error(f"{err.message} ({err.path})", code="INPUT_ERROR")

    # --- Build request --------------------------------------------------
    # Import the SDK lazily so capability/input errors don't pay the import cost.
    from google import genai
    from google.genai import types

    full_prompt = (
        f"{prompt}. Do not include: {negative_prompt}." if negative_prompt else prompt
    )

    parts: list[types.Part] = [types.Part(text=full_prompt)]
    for img in input_images:
        parts.append(
            types.Part(
                inline_data=types.Blob(
                    mime_type=img.mime_type,
                    data=img.data_bytes,
                )
            )
        )

    config_kwargs: dict[str, Any] = {"response_modalities": ["IMAGE"]}
    image_config_kwargs: dict[str, Any] = {}
    if aspect_ratio:
        image_config_kwargs["aspect_ratio"] = aspect_ratio
    if size:
        image_config_kwargs["image_size"] = size
    if image_config_kwargs:
        config_kwargs["image_config"] = types.ImageConfig(**image_config_kwargs)
    if thinking:
        # SDK enum is upper-case (MINIMAL | HIGH). Normalize on the way out.
        config_kwargs["thinking_config"] = types.ThinkingConfig(
            thinking_level=thinking.upper()
        )
    config = types.GenerateContentConfig(**config_kwargs)

    if output:
        output_path = Path(output).resolve()
    else:
        # Date.now() in TS gives ms; preserve millisecond precision.
        output_path = (Path.cwd() / f"generated_{int(time.time() * 1000)}.png").resolve()

    # --- Call API -------------------------------------------------------
    try:
        client = genai.Client(api_key=resolved.key)
        response = client.models.generate_content(
            model=model,
            contents=[types.Content(role="user", parts=parts)],
            config=config,
        )

        response_parts = None
        if response.candidates and response.candidates[0].content:
            response_parts = response.candidates[0].content.parts
        if not response_parts:
            json_error("No content in API response", code="API_ERROR")

        image_parts = [
            p
            for p in response_parts
            if p.inline_data
            and p.inline_data.mime_type
            and p.inline_data.mime_type.startswith("image/")
        ]
        if not image_parts:
            json_error("API response did not contain an image", code="API_ERROR")

        non_thought = [p for p in image_parts if not getattr(p, "thought", False)]
        final = (non_thought or image_parts)[-1]
        blob = final.inline_data
        if not blob or not blob.data or not blob.mime_type:
            json_error("Image data missing from API response", code="API_ERROR")

        # SDK returns raw bytes for inline_data.data; tolerate the rare base64
        # string shape some versions have produced.
        raw = blob.data
        if isinstance(raw, str):
            import base64

            data_bytes = base64.b64decode(raw)
        else:
            data_bytes = raw

        try:
            output_path.parent.mkdir(parents=True, exist_ok=True)
            output_path.write_bytes(data_bytes)
        except OSError as exc:
            json_error(
                f"Could not write output file: {exc}", code="OUTPUT_ERROR"
            )

        result: dict[str, Any] = {
            "file": str(output_path),
            "mimeType": blob.mime_type,
            "size": len(data_bytes),
            "prompt": prompt,
            "model": model,
        }
        if aspect_ratio:
            result["aspectRatio"] = aspect_ratio
        if size:
            result["imageSize"] = size
        if thinking:
            result["thinkingLevel"] = thinking
        if input_images:
            result["inputImages"] = [str(i.absolute_path) for i in input_images]
        json_ok(**result)
    except SystemExit:
        # Re-raise json_ok / json_error sentinels so they aren't swallowed.
        raise
    except Exception as exc:  # noqa: BLE001 — match TS: fold every API failure under API_ERROR
        message = str(exc) or "Unknown API error"
        json_error(message, code="API_ERROR")
